pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -199,6 +199,7 @@ public boolean isMultiBatches() {
     public OpSendMsg createOpSendMsg() throws IOException {
         ByteBuf encryptedPayload = producer.encryptMessage(messageMetadata, getCompressedBatchMetadataAndPayload());
         if (encryptedPayload.readableBytes() > ClientCnx.getMaxMessageSize()) {
+            messages.forEach(msg -> producer.client.getMemoryLimitController().releaseMemory(msg.getUncompressedSize()));
```

Comment:
Do we need release semaphore here?

Reply:
@gaozhangmin no need to release semaphore. I have created a testcase to cover this.

Reply:
Why? Could u make some explain?